controllers/bootstrap/init.go

```diff
@@ -172,21 +173,71 @@ func (b *Bootstrap) CrossplaneCloudOperator(instance *apiv3.CommonService) error
 	}
 
 	if bedrockshim {
+		if b.CSData.CrossplaneEnabledCR == nil {
+			b.CSData.CrossplaneEnabledCR = make(map[string]bool)
+		}
+		b.CSData.CrossplaneEnabledCR[instance.Namespace+"/"+instance.Name] = true
+
 		b.CSData.CrossplaneProvider = "odlm"
 
 		if b.SaasEnable {
 			b.CSData.CrossplaneProvider = "ibmcloud"
 			if err := b.installCloudOperator(); err != nil {
 				return err
 			}
-
 		}
 
 		if err := b.installCrossplaneOperator(); err != nil {
 			return err
 		}
+	} else {
+		if err := b.DeleteCrossplaneCloudSubscription(b.CSData.MasterNs, instance.Namespace+"/"+instance.Name); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DeleteCrossplaneCloudSubscription deleted crossplane & cloud operator subscription when bedrockshim set to false or CS CR is removed
+func (b *Bootstrap) DeleteCrossplaneCloudSubscription(namespace, NamespacedName string) error {
+	if b.CSData.CrossplaneEnabledCR[NamespacedName] {
+		delete(b.CSData.CrossplaneEnabledCR, NamespacedName)
+	} else {
+		return nil
```

Comment:
I am wondering that if we remove `bedrockshim ` from CR and restart the cs-operator at the same time, then because of the `CrossplaneEnabledCR` is empty, it seems that we could never delete the Crossplane.

But if we remove the `else, return nil`. I am worried that if we have two CRs which one has `bedrockshim` and the other does not have `bedrockshim`.
After restarting cs-operator, if the CR without `bedrockshim` is reconciled first, then the crossplane will be removed and re-installed when reconciling the other CR.

Reply:
I think the key to solving this problem is how to get all the `bedrockshim` values from all cs cr after restart.  There are two possible ways come. 
1. could we save the `b.CSData.CrossplaneEnabledCR` in a permanent place? such as write to file or some permanent memory
2. after the cs-operator restarted, it will re-generate all cs cr. Do we know how many cs cr will be re-generated OR what cs cr will be re-generated? (I think we know since the new generated cs cr will have the same data as before). could we fetch each `bedrockshim` value of all cs cr when we re-generate them?

Reply:
> Could we save the b.CSData.CrossplaneEnabledCR in a permanent place? such as write to file or some permanent memory

Technically, we could store the information into the ConfigMap, and read it when we need the data.

> after the cs-operator restarted, it will re-generate all cs cr. Do we know how many cs cr will be re-generated OR what cs cr will be re-generated? (I think we know since the new generated cs cr will have the same data as before). could we fetch each bedrockshim value of all cs cr when we re-generate them?

A possible solution is that we go through all the CS CRs when those resources are determined by all the CS CRs in the cluster.
In this case, we don't need to worry about whether the operator will be restarted, because no matter which CR is reconciled first, we always fetch every `bedrockshim` values to determine the uninstallation.

We have a example here when we merge the size configuration across the CRs to OperandConfig.

When there is one CS CR deleted in the cluster, we have the function called `handleDelete()`
https://github.com/IBM/ibm-common-service-operator/blob/master/controllers/operandconfig.go#L364
After the deletion, the size configuration in that CR should not apply to OperandConfig anymore.

Therefore, we fetch all the CRs which remain in the cluster, then get a new set of configuration from remaining CR and apply it to the OperandConfig.
https://github.com/IBM/ibm-common-service-operator/blob/36da1ebbc69d8450629463b7ffecee1789ee1dda/controllers/operandconfig.go#L318

I think we could apply to same logic to the uninstallation of crossplane.

When a CR is deleted or its `bedrockshim` value is removed, we fetch remaining CRs and determined whether the crossplane should keep in the cluster or not.